### PR TITLE
Allow trainer to return eval. loss for CLIP-like models

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

Allow trainer to give **evaluation** loss for CLIP-like models.

Currently, this line
https://github.com/huggingface/transformers/blob/07d8d6e2f7a920d399e5e86a82d78179cdfa6746/src/transformers/trainer.py#L3192
gives `has_labels = False` for CLIP-like models, and can't give loss value in the evaluation.

without this PR:
```bash
***** eval metrics *****
  epoch                   =        1.0
  eval_runtime            = 0:00:01.67
  eval_samples_per_second =      9.571
  eval_steps_per_second   =      4.785

```

with this PR.
```bash
***** eval metrics *****
  epoch                   =        1.0
  eval_loss               =     0.8159
  eval_runtime            = 0:00:01.66
  eval_samples_per_second =      9.598
  eval_steps_per_second   =      4.799

```

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -3190,6 +3192,9 @@ def prediction_step(
             logits and labels (each being optional).
         """
         has_labels = False if len(self.label_names) == 0 else all(inputs.get(k) is not None for k in self.label_names)
+        # For CLIP-like models capable of returning loss values.
+        can_compute_loss = True if len(self.label_names) == 0 and self.can_return_loss else False
```
Comment (ydshieh) [inline src/transformers/trainer.py:3196]: We need to restrict to `len(self.label_names) == 0`.

For models that has `len(self.label_names) > 0`, we should check if the inputs contain the required labels by the model - which is done one line above `has_labels`.

Comment (ydshieh) [reply src/transformers/trainer.py:3196]: `can_compute_loss` is actually `can_compute_loss_without_labels`, but maybe a too long name.

Comment (sgugger) [reply src/transformers/trainer.py:3196]: Can we rename all `can_compute_loss` to `loss_without_labels` then? It's more informative even if not completely perfect.

File: src/transformers/utils/generic.py [hunk 2]
```diff
@@ -336,6 +336,23 @@ def __exit__(self, *args, **kwargs):
...
+        model_class (`type`): The class of the model.
+    """
+    model_name = model_class.__name__
+    if model_name.startswith("TF"):
+        signature = inspect.signature(model_class.call)
+    elif model_name.startswith("Flax"):
+        signature = inspect.signature(model_class.__call__)
+    else:
+        signature = inspect.signature(model_class.forward)
+    return [p for p in signature.parameters if p == "return_loss"]
```
Comment (sgugger) [inline src/transformers/utils/generic.py:353]: I'd prefer if this returns a `bool` then rely on Python magic conversion to bools.
Also I think we should check if the default is `True` as the Trainer won't change the default.

**Review by sgugger — COMMENTED** [review]
Very nice idea thanks a lot! I left a few comments regarding naming and we can make the function `can_return_loss` a bit better, but overall great PR!

File: src/transformers/utils/generic.py (hunk 2 above)
Comment (ydshieh) [reply src/transformers/utils/generic.py]: I changed it to 
```python
any(p == "return_loss" for p in signature.parameters)
```
but do you mean we should have sth. like (conceptually)

```pythont 
any(p == "return_loss" and default_value(p) is True for p in signature.parameters)
```
?

Comment (sgugger) [reply src/transformers/utils/generic.py]: Yes to the second!

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -3190,6 +3192,9 @@ def prediction_step(
             logits and labels (each being optional).
         """
         has_labels = False if len(self.label_names) == 0 else all(inputs.get(k) is not None for k in self.label_names)
+        # For CLIP-like models capable of returning loss values.
+        loss_without_labels = True if len(self.label_names) == 0 and self.can_return_loss and inputs.get("return_loss", None) is True else False
```
[the remark this answers is not in the document]
Comment (ydshieh) [reply src/transformers/trainer.py:3196]: OK! I see all of our `return_loss` have `None` as default value (include `CLIP`), but I can add extra check to be sure

[the remark this answers is not in the document]
Comment (sgugger) [reply src/transformers/trainer.py:3196]: Oh in this case then your solution works.

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -3193,7 +3193,12 @@ def prediction_step(
         """
         has_labels = False if len(self.label_names) == 0 else all(inputs.get(k) is not None for k in self.label_names)
         # For CLIP-like models capable of returning loss values.
-        loss_without_labels = True if len(self.label_names) == 0 and self.can_return_loss and inputs.get("return_loss", None) is True else False
+        # If `return_loss` is not specified or being `None` in `inputs`, we check if the default value of `return_loss`
+        # is `True` in `model.forward`.
+        return_loss = inputs.get("return_loss", None)
+        if return_loss is None:
```
Comment (ydshieh) [inline src/transformers/trainer.py:3199]: If `False`, no need to check default value

File: src/transformers/utils/generic.py [hunk 5]
```diff
@@ -336,6 +336,28 @@ def __exit__(self, *args, **kwargs):
...
+    if model_name.startswith("TF"):
+        signature = inspect.signature(model_class.call)
+    elif model_name.startswith("Flax"):
+        signature = inspect.signature(model_class.__call__)
+    else:
+        signature = inspect.signature(model_class.forward)
+
+    for p in signature.parameters:
+        if p == "return_loss" and signature.parameters[p].default is True:
+            return True
```
Comment (ydshieh) [inline src/transformers/utils/generic.py:356]: Will return `True` only if the default value is `True`.

**Comment by ydshieh** [comment]
@sgugger Hopefully the change covers everything that could happen now and in the future.

**Review by sgugger — APPROVED** [review]
Perfect!
